lib/transform/classic.js

```diff
@@ -0,0 +1,194 @@
+    .get('declaration', 'arguments')
+    .filter(path => path.value.type === 'ObjectExpression');
+
+  let objectArg = extendObjectArgs[0];
+  if (!objectArg) {
+    throw new SilentError(
+      `Could not find object argument in \`export default Component.extend({ ... });\``
+    );
+  }
+
+  // find `tagName` property if it exists
+  let properties = objectArg.get('properties');
+  let tagName = findTagName(properties);
+
+  // skip tagless components (silent)
+  if (tagName === '') {
+    debug('tagName: %o -> skip', tagName);
+    return;
+  }
+
+  debug('tagName: %o', tagName);
+
+  // skip components that use `this.element`
+  let thisElementPaths = j(objectArg).find(j.MemberExpression, {
+    object: { type: 'ThisExpression' },
+    property: { name: 'element' },
+  });
+  if (thisElementPaths.length !== 0) {
+    throw new SilentError(`Using \`this.element\` is not supported in tagless components`);
+  }
+
+  // skip components that use `this.elementId`
+  let thisElementIdPaths = j(objectArg).find(j.MemberExpression, {
+    object: { type: 'ThisExpression' },
+    property: { name: 'elementId' },
+  });
+  if (thisElementIdPaths.length !== 0) {
+    throw new SilentError(`Using \`this.elementId\` is not supported in tagless components`);
+  }
+
+  // skip components that use `click()` etc.
+  for (let methodName of EVENT_HANDLER_METHODS) {
+    let handlerMethod = properties.filter(path => isMethod(path, methodName))[0];
+    if (handlerMethod) {
+      throw new SilentError(`Using \`${methodName}()\` is not supported in tagless components`);
+    }
+  }
+
+  // analyze `elementId`, `attributeBindings`, `classNames` and `classNameBindings`
+  let elementId = findElementId(properties);
+  debug('elementId: %o', elementId);
+
+  let attributeBindings = findAttributeBindings(properties);
+  debug('attributeBindings: %o', attributeBindings);
+
+  let classNames = findClassNames(properties);
+  debug('classNames: %o', classNames);
+
+  let classNameBindings = findClassNameBindings(properties);
+  debug('classNameBindings: %o', classNameBindings);
+
+  // set `tagName: ''`
+  let tagNamePath = j(properties)
+    .find(j.ObjectProperty)
+    .filter(path => path.parentPath === properties)
+    .filter(path => isProperty(path, 'tagName'));
+
+  if (tagNamePath.length === 1) {
+    j(tagNamePath.get('value')).replaceWith(j.stringLiteral(''));
+  } else {
+    properties.unshift(j.objectProperty(j.identifier('tagName'), j.stringLiteral('')));
+  }
+
+  // remove `elementId`, `attributeBindings`, `classNames` and `classNameBindings`
+  j(properties)
+    .find(j.ObjectProperty)
+    .filter(path => path.parentPath === properties)
+    .filter(
+      path =>
+        isProperty(path, 'elementId') ||
+        isProperty(path, 'attributeBindings') ||
+        isProperty(path, 'classNames') ||
+        isProperty(path, 'classNameBindings')
+    )
+    .remove();
+
+  let newSource = root.toSource();
+
+  // wrap existing template with root element
```

Comment:
I'm wondering if everything below should rather be in the `transformTemplate()` function instead 🤔

Reply:
Makes sense! Keeps the template AST builder stuff isolated in that transform.